### PR TITLE
Add toMap keyword to create homogeneous maps from records

### PR DESCRIPTION
**Pull request by blamario**
The tests are rather rudimentary, let me know if/what more is required.

### CODE REVIEW

**Comment by Gabriel439** [comment]
@blamario: We need this to be standardized according to https://github.com/dhall-lang/dhall-lang/blob/master/standard/README.md, but I can help with that

**Comment by blamario** [comment]
I won't be able to do anything until Tuesday. Should it be done before the merge?

**Comment by Gabriel439** [comment]
Yes.  The reason why is that the feature still needs to go through an official proposal process.  See: https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#how-do-i-change-the-language

File: tests/normalization/success/unit/EmptyToMapA.dhall [hunk 1]
```diff
@@ -0,0 +1 @@
+toMap {=} : List {mapKey : Text, mapValue : Natural}
```
Comment (singpolyma) [inline tests/normalization/success/unit/EmptyToMapA.dhall]: missing newline

File: tests/type-inference/success/unit/ToMapB.dhall [hunk 2]
```diff
@@ -0,0 +1 @@
+List { mapKey : Text, mapValue : Natural }
```
Comment (singpolyma) [inline tests/type-inference/success/unit/ToMapB.dhall]: missing newline

**Comment by blamario** [comment]
I have modified the standard description with the last commit, please review.

One of my tests is failing:

```
./dhall-lang/tests/normalization/success/unit/ToMapWithType:                              FAIL
        tests/Dhall/Test/Normalization.hs:180:
        The normalized expression did not match the expected output
        expected: Annot (ListLit Nothing (fromList [RecordLit (fromList [("mapKey",TextLit (Chunks [] "bar")),("mapValue",NaturalLit 4)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "baz")),("mapValue",NaturalLit 9)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "foo")),("mapValue",NaturalLit 1)])])) (App List (Record (fromList [("mapKey",Text),("mapValue",Natural)])))
         but got: ListLit (Just (Record (fromList [("mapKey",Text),("mapValue",Natural)]))) (fromList [RecordLit (fromList [("mapKey",TextLit (Chunks [] "bar")),("mapValue",NaturalLit 4)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "baz")),("mapValue",NaturalLit 9)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "foo")),("mapValue",NaturalLit 1)])])

**Comment by sjakobi** [comment]
Can you add some tests of expected failures? I can only think of inhomogenous records right now, but maybe there are other scenarios.

**Comment by sjakobi** [comment]
> Can you add some tests of expected failures? I can only think of inhomogenous records right now, but maybe there are other scenarios.

What about `toMap {=}` without a type annotation?

Did you consider adding an argument for the type of `mapValue` BTW?

**Comment by blamario** [comment]
Ok, I added several failure tests. I stuffed them all into `tests/typecheck/failure/unit/` directory, let me know if any belong elsewhere.

> Did you consider adding an argument for the type of `mapValue` BTW?

I didn't. Its addition wouldn't make `toMap` typeable, so what would be the point?

**Comment by sjakobi** [comment]
> Ok, I added several failure tests. I stuffed them all into `tests/typecheck/failure/unit/` directory, let me know if any belong elsewhere.

Thank you! :+1: 

> > Did you consider adding an argument for the type of `mapValue` BTW?
> 
> I didn't. Its addition wouldn't make `toMap` typeable, so what would be the point?

E.g. `toMap Natural {=}` could be allowed then, I believe. But yeah, that's not a great win. ;)

**Comment by blamario** [comment]
What would have been a win, in some situations at least, is if a homogeneous record type with field value type `A` could be defined as a map from an enumeration type to `A`. For example,

```
let Enum = <foo | bar> in
let r = {foo : 3, bar : 8} : Enum -> Natural in
toMap r : List {mapKey : Enum, mapValue : Natural}
```

However this would have a much wider impact.

File: standard/type-inference.md [hunk 3]
```diff
@@ -753,6 +753,22 @@ If the operator arguments are not record types then that is a type error.
 If they share a field in common that is not a record type then that is a type
 error.
 
+
+
+The `toMap` operator cab be applied only to a record value, and every field
+of the record must have the same type.
+
+
+    Γ ⊢ e :⇥ { x : T, xs… }
```
Comment (sjakobi) [inline standard/type-inference.md]:
```suggestion
    Γ ⊢ e :⇥ { x : T₀, xs… }
    Γ ⊢ toMap { xs… } :⇥ List { mapKey : Text, mapValue : T₁ }
    T₀ ≡ T₁
```

File: standard/type-inference.md [hunk 4]
```diff
@@ -753,6 +753,22 @@ If the operator arguments are not record types then that is a type error.
...
+    ──────────────────────────────────────────────────
+    Γ ⊢ toMap e : List { mapKey : Text, mapValue : T }
```
Comment (sjakobi) [inline standard/type-inference.md]:
```suggestion
    Γ ⊢ toMap e : List { mapKey : Text, mapValue : T₀ }
```

File: standard/type-inference.md [hunk 5]
```diff
@@ -753,6 +753,22 @@ If the operator arguments are not record types then that is a type error.
...
+
+
+    Γ ⊢ e :⇥ {}    Γ ⊢ T :⇥ Type
```
Comment (sjakobi) [inline standard/type-inference.md]:
```suggestion
    Γ ⊢ e :⇥ {}    Γ ⊢ T₀ :⇥ List { mapKey : Text, mapValue : T₁ }
```

File: standard/type-inference.md [hunk 6]
```diff
@@ -753,6 +753,22 @@ If the operator arguments are not record types then that is a type error.
...
+    ────────────────────────────
+    Γ ⊢ ( toMap e : T ) : T
```
Comment (sjakobi) [inline standard/type-inference.md]:
```suggestion
    Γ ⊢ ( toMap e : T₀ ) : T₀
```

Comment (sjakobi) [reply standard/type-inference.md]: Maybe that's better:

```suggestion
    Γ ⊢ ( toMap e : T₀ ) : List { mapKey : Text, mapValue : T₁ }
```

File: standard/type-inference.md (hunk 3 above)
Comment (blamario) [reply standard/type-inference.md]: I don't know enough about the logic of inference rules to accept or reject this.

File: standard/type-inference.md (hunk 4 above)
Comment (blamario) [reply standard/type-inference.md]: Wouldn't `T₀` be unbound here? Did you mean to rename the `T` to `T₀` as well?

Comment (sjakobi) [reply standard/type-inference.md]: Ah sorry, I should have made a proper pull request. This change belongs to the other one above the line.

File: standard/type-inference.md (hunk 3 above)
Comment (Gabriel439) [reply standard/type-inference.md]: @blamario: The proposed change looks correct to me.  It's saying that each field of the record passed to `toMap` must have the same type

**Comment by blamario** [comment]
Ok, I've incorporated the suggestions. Do you want me to update the fork before the merge?

**Comment by Gabriel439** [comment]
@blamario: Wait up to 4 more days for @f-f and @singpolyma to review since the standard-related changes were ready 3 days ago.  For more details, see: https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#how-do-changes-get-approved

File: standard/beta-normalization.md [hunk 7]
```diff
@@ -1069,6 +1069,41 @@ record types:
     l₀ ⩓ r₀ ⇥ l₁ ⩓ r₁
 
 
+A record whose fields all have the same type (*i.e.*, a *homogenous* record) can be converted to a list where each list
```
Comment (f-f) [inline standard/beta-normalization.md]:
```suggestion
A record whose fields all have the same type (*i.e.*, a *homogeneous* record) can be converted to a list where each list
```

File: standard/dhall.abnf [hunk 8]
```diff
@@ -600,36 +601,42 @@ import = import-hashed [ whsp as whsp1 (Text / Location) ]
 expression =
     ; "\(x : a) -> b"
       lambda whsp "(" whsp nonreserved-label whsp ":" whsp1 expression whsp ")" whsp arrow whsp expression
-    
+
```
Comment (f-f) [inline standard/dhall.abnf]: I think we should not remove whitespace here in the grammar, see #379

File: standard/binary.md [hunk 9]
```diff
@@ -460,6 +460,22 @@ have a type annotation:
     encode(merge t₀ u₀ : T₀) = [ 6, t₁, u₁, T₁ ]
 
 
+### `toMap` expressions
+
+`toMap` expressions differ in their encoding depending on whether or not they
```
Comment (singpolyma) [inline standard/binary.md]: While I don't mind this encoding, it is normal throughout to use `null` for a missing type annotation.

File: standard/binary.md [hunk 10]
```diff
@@ -652,7 +668,7 @@ interpolated expressions:
     encode("a${b₀}c${d}e…x${y₀}z") = [ 18, "a", b₁, "c", d₁, "e", …, "x", y₁, "z" ]
 
 In other words: the amount of encoded elements is always an odd number, with the
-odd elements being strings and the even ones being interpolated expressions.
```
Comment (singpolyma) [inline standard/binary.md]: Unrelated whitespace change.

File: standard/type-inference.md [hunk 11]
```diff
@@ -753,6 +753,24 @@ If the operator arguments are not record types then that is a type error.
 If they share a field in common that is not a record type then that is a type
 error.
 
+
+
+The `toMap` operator cab be applied only to a record value, and every field
```
Comment (singpolyma) [inline standard/type-inference.md]: s/cab/can

**Comment by singpolyma** [comment]
Question: should we consider using an enum for the generated keys instead of a Text?  For example:

```dhall
toMap { a = 1, b = 2 } -> [{ mapKey = <a|b>.a, mapValue = 1 }, { mapKey = <a|b>.b, mapValue = 2 } ]
```

This is easy to convert to the `Text` version by mapping `merge` if needed, but since Text is mostly opaque this seems like it would be more expressive/flexible while also more closely representing the "atoms" that record selectors are.

File: standard/dhall.abnf (hunk 8 above)
Comment (blamario) [reply standard/dhall.abnf]: My Emacs was set up to remove all trailing whitespace on save, but I can restore it.

File: standard/binary.md (hunk 9 above)
Comment (blamario) [reply standard/binary.md]: I was following the `merge` encoding, which doesn't use `null` either.

**Comment by Gabriel439** [comment]
@singpolyma: My understanding is that the main motivation for `toMap` is to provide an ergonomic way to generate association lists, but if they have to be post-processed with a `List/map` then they lose most of the ergonomic benefit.

For example, the current proposal plays well with the recent addition to the Prelude of weakly-typed JSON support (which expects `Text` keys), like this:

```dhall
let JSON = https://prelude.dhall-lang.org/JSON/package.dhall

in  JSON.object (toMap { foo = JSON.null, bar = JSON.number 1.0 })
```

Similarly, the current proposal would also play nice with your proposal to use `Prelude/Map` for custom headers (https://github.com/dhall-lang/dhall-lang/pull/611):

```dhall
https://example.com/foo.dhall using toMap { Authorization = "token ${env:TOKEN as Text }" }
```

If you change both examples to use unions for keys then they become much more verbose, in which case they might as well have been written out as explicit association lists.

**Comment by Gabriel439** [comment]
@blamario: You have the necessary approvals, so you can merge once you fix the CI failures (which looks like a missing trailing newline): https://hydra.dhall-lang.org/build/20574/nixlog/1/tail

File: standard/beta-normalization.md [hunk 12]
```diff
@@ -1066,6 +1066,41 @@ record types:
...
+    ──────────────────────────────────────────────
+    toMap t ⇥ [ {mapKey = "x", mapValue = v} ] # m
+
+
+The `toMap` application can be annotated with a type, and it must be if the record is empty.
+
+
+    t ⇥ { x = v, ts… }   toMap { ts } ⇥ m   T₀ ⇥ T₁
+    ────────────────────────────────────────────────────────────
+    toMap t : T₀ ⇥ ( [ {mapKey = "x", mapValue = v} ] # m ) : T₁
```
Comment (Gabriel439) [inline standard/beta-normalization.md]: The CI failure for the matching Haskell pull request caught an issue here: a non-empty list with a type annotation is not in normal form:

```suggestion
    toMap t : T₀ ⇥ [ {mapKey = "x", mapValue = v} ] # m
```

This will also require a matching change to the `./tests/normalization/success/unit/ToMapWithType` test

File: standard/beta-normalization.md [hunk 13]
```diff
@@ -1066,6 +1066,41 @@ record types:
...
+
+    t ⇥ { x = v, ts… }   toMap { ts } ⇥ m
+    ──────────────────────────────────────────────
+    toMap t ⇥ [ {mapKey = "x", mapValue = v} ] # m
+
+
+The `toMap` application can be annotated with a type, and it must be if the record is empty.
+
+
+    t ⇥ { x = v, ts… }   toMap { ts } ⇥ m   T₀ ⇥ T₁
```
Comment (Gabriel439) [inline standard/beta-normalization.md]: See the following comment

```suggestion
    t ⇥ { x = v, ts… }   toMap { ts } ⇥ m
```

File: standard/binary.md (hunk 10 above)
Comment (singpolyma) [reply standard/binary.md]: This one looks like it's still here.

File: standard/dhall.abnf [hunk 14]
```diff
@@ -575,7 +576,7 @@ bash-environment-variable = (ALPHA / "_") *(ALPHA / DIGIT / "_")
 ; character like `env`
 posix-environment-variable = 1*posix-environment-variable-character
 
-; These are all the characters from the POSIX Portable Character Set except for
```
Comment (singpolyma) [inline standard/dhall.abnf]: unrelated whitespace change

File: standard/shift.md [hunk 15]
```diff
@@ -539,4 +549,3 @@ The remaining rules are:
 
     ───────────────────────
     ↑(d, x, m, Sort) = Sort
-
```
Comment (singpolyma) [inline standard/shift.md]: unrelated whitespace change

File: standard/type-inference.md [hunk 16]
```diff
@@ -753,6 +753,24 @@ If the operator arguments are not record types then that is a type error.
...
+
+    Γ ⊢ e :⇥ { x : T₀, xs… }
+    Γ ⊢ toMap { xs… } :⇥ List { mapKey : Text, mapValue : T₁ }
+    T₀ ≡ T₁
+    ──────────────────────────────────────────────────
+    Γ ⊢ toMap e : List { mapKey : Text, mapValue : T₀ }
+
+
+    Γ ⊢ e :⇥ {}    Γ ⊢ T₀ :⇥ List { mapKey : Text, mapValue : T₁ }
+    ────────────────────────────
```
Comment (singpolyma) [inline standard/type-inference.md]: neither this nor the previous line is long enough

File: standard/beta-normalization.md (hunk 12 above)
Comment (blamario) [reply standard/beta-normalization.md]: I suppose this is also the reason for the failure of `tests/normalization/success/unit/ToMapWithType` I get in `dhall-haskell`:

      ./dhall-lang/tests/normalization/success/unit/ToMapWithType:                              FAIL
        tests/Dhall/Test/Normalization.hs:176:
        The normalized expression did not match the expected output
        expected: Annot (ListLit Nothing (fromList [RecordLit (fromList [("mapKey",TextLit (Chunks [] "bar")),("mapValue",NaturalLit 4)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "baz")),("mapValue",NaturalLit 9)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "foo")),("mapValue",NaturalLit 1)])])) (App List (Record (fromList [("mapKey",Text),("mapValue",Natural)])))
         but got: ListLit (Just (Record (fromList [("mapKey",Text),("mapValue",Natural)]))) (fromList [RecordLit (fromList [("mapKey",TextLit (Chunks [] "bar")),("mapValue",NaturalLit 4)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "baz")),("mapValue",NaturalLit 9)]),RecordLit (fromList [("mapKey",TextLit (Chunks [] "foo")),("mapValue",NaturalLit 1)])])

I have to ask again: if the normal form of `[4] : List Natural` is *not* supposed to be `ListLit (Just Natural) (pure $ NaturalLit 4)` why does the `ListLit` constructor even allow a type annotation on a non-empty list? This is misleading. It would be better to have `ListLit (Either (Expr s a) (NonEmpty (Expr s a))` to avoid the ambiguity.

**Comment by blamario** [comment]
> @blamario: You have the necessary approvals, so you can merge once you fix the CI failures (which looks like a missing trailing newline): https://hydra.dhall-lang.org/build/20574/nixlog/1/tail

If you mean merge the PR, then no I can't. I don't have the write access to the `dhall-lang` repository.

**Comment by Gabriel439** [comment]
@blamario: Whoops!  I forgot to add you as a contributor.  I just sent you an invite so that you can merge when ready
